programs/ssl/ssl_test_lib.c

```diff
@@ -193,6 +193,34 @@ int rng_get( void *p_rng, unsigned char *output, size_t output_len )
 #endif /* !MBEDTLS_TEST_USE_PSA_CRYPTO_RNG */
 }
 
+int key_opaque_alg_parse( const char *arg, const char **alg1, const char **alg2 )
+{
+    char* separator;
+    if( ( separator = strchr( arg, ',' ) ) == NULL )
+        return 1;
```

Comment:
This means args must contain at at least 2 entries, this should either be added in the `key_opaque_algs` description or use `none` for the second alg.

Reply:
> Add a new command-line option key_opaque_algs to ssl_server2 and ssl_client2, that accepts a coma-separated pair of values among the following:
"rsa-sign-pkcs1"
"rsa-sign-pss"
"rsa-decrypt"
"ecdsa-sign"
"ecdh"
"none" (only acceptable for the second value).

Information that none is designed only for second value can be found here:
https://github.com/Mbed-TLS/mbedtls/pull/5779/files#diff-323c6732e9483b4898e122ef59a2fe4e05be967bb426760fbbe241933348a95bR351-R352

Reply:
Something like this would meet the documented behavior:
```c
int key_opaque_alg_parse( const char *arg, const char **alg1, const char **alg2 )
{
    char* arg2;
   
    arg2 = strchr( arg, ',' );

    if( arg2 != NULL )
    {
        *arg2 = '\0';
        *arg2 += 1;
    }

    if( strcmp( *arg, "rsa-sign-pkcs1" ) != 0 &&
        strcmp( *arg, "rsa-sign-pss" ) != 0 &&
        strcmp( *arg, "rsa-decrypt" ) != 0 &&
        strcmp( *arg, "ecdsa-sign" ) != 0 &&
        strcmp( *arg, "ecdh" ) != 0 )
        return 1;

    *alg1 = arg;

    /* Return and leave alg2 to default "none" value */
    if( arg2 == NULL )
        return 0;

    if( strcmp( *arg2, "rsa-sign-pkcs1" ) != 0 &&
        strcmp( *arg2, "rsa-sign-pss" ) != 0 &&
        strcmp( *arg2, "rsa-decrypt" ) != 0 &&
        strcmp( *arg2, "ecdsa-sign" ) != 0 &&
        strcmp( *arg2, "ecdh" ) != 0 &&
        strcmp( *arg2, "none" ) != 0 )
        return 1;

    *alg1 = arg2;

    return 0;
}
```

Reply:
The documentation and #5763 implies all these should be valid:
```
key_opaque_algs=rsa-sign-pkcs1
key_opaque_algs=rsa-sign-pkcs1,rsa-sign-pss
key_opaque_algs=rsa-sign-pkcs1,rsa-sign-pss,rsa-decrypt
key_opaque_algs=rsa-sign-pkcs1,rsa-sign-pss,rsa-decrypt,ecdsa-sign
key_opaque_algs=rsa-sign-pkcs1,rsa-sign-pss,rsa-decrypt,ecdsa-sign,ecdh
key_opaque_algs=rsa-sign-pkcs1,none
key_opaque_algs=rsa-sign-pkcs1,rsa-sign-pss,none
key_opaque_algs=rsa-sign-pkcs1,rsa-sign-pss,rsa-decrypt,none
key_opaque_algs=rsa-sign-pkcs1,rsa-sign-pss,rsa-decrypt,ecdsa-sign,none
key_opaque_algs=rsa-sign-pkcs1,rsa-sign-pss,rsa-decrypt,ecdsa-sign,ecdh,none
```

but
```
key_opaque_algs=
```
and
```
key_opaque_algs=none
```

should be invalid

Reply:
Only pair of of values is valid, so why
```
key_opaque_algs=rsa-sign-pkcs1,rsa-sign-pss,rsa-decrypt
key_opaque_algs=rsa-sign-pkcs1,rsa-sign-pss,rsa-decrypt,ecdsa-sign
key_opaque_algs=rsa-sign-pkcs1,rsa-sign-pss,rsa-decrypt,ecdsa-sign,ecdh
```

this should be valid?

I think we should check that we have no comma:
`key_opaque_algs=rsa-sign-pkcs1`
or up to one comma:
`key_opaque_algs=rsa-sign-pkcs1,rsa-sign-pss`

Reply:
@mprse you're right for the pair, so not sure about the validity of a single value with no comma

Reply:
Note: the purpose is not just to parse things for the fun of it, but to eventually pass it to `mbedtls_pk_wrap_as_opaque()`, which (once #5761 is done) can only set one or two algs, no less, no more.

Now, the question is whether the one alg case can be written `key_opaque_algs=rsa-sign-pkcs1` or it we require people to write it as `key_opaque_algs=rsa-sign-pkcs1,none`. I think allowing the first form (as equivalent to the second) would be user-friendly, at the code of making the parsing cost a bit more complex. I think it's up to you to implement it or not. (Users of this option would be mostly us, for testing, so we can decide to be less user-friendly if we want.)